src/Simulation/Common/QubitManager.cs

```diff
@@ -17,22 +16,35 @@ namespace Microsoft.Quantum.Simulation.Common
     /// Allocation and release are O(1) operations.
     /// QubitManager uses memory - sizeof(long)*qubitCapacity. qubitCapacity is passed in to the constructor.
     /// </summary>
-    public class QubitManager
+    public class QubitManager : IQubitManager
     {
-        long NumQubits; // Total number of qubits this QubitManager is capable of handling.
-        long None; // Indicates the end of the list of free qubits. If we reach it - we are out of qubits.
-        long Allocated; // All qubits allocated by user will be marked with this number.
-        long Disabled; // All qubits disabled via DisableQubit interface will be marked with this number.
-        long AllocatedForBorrowing; // All qubits allocated only for borrowing, will be marked with this number or higher.
-        long[] qubits; // Tracks the allocation state of all qubits.
-        long free; // Points to the first free (unallocated) qubit.
-        long freeTail; // Points to the last free (unallocated) qubit. Only valid iff (!EncourageReuse).
-        long numAllocatedQubits; // Tracking this for optimization.
-        long numDisabledQubits; // Number of disabled qubits.
+        /// <summary>
+        /// Total number of qubits this QubitManager is capable of handling.
+        /// </summary>
+        protected long NumQubits;
+        /// <summary>
+        /// The number of currently allocated qubits.
+        /// </summary>
+        protected long NumAllocatedQubits;
+        /// <summary>
+        /// The number of disabled qubits.
+        /// </summary>
+        protected long NumDisabledQubits;
+        /// <summary>
+        /// Tracks the allocation state of all qubits.
+        /// </summary>
+        protected long[] qubits; 
+
+        private long None; // Indicates the end of the list of free qubits. If we reach it - we are out of qubits.
+        private long Allocated; // All qubits allocated by user will be marked with this number.
+        private long Disabled; // All qubits disabled via DisableQubit interface will be marked with this number.
+        private long AllocatedForBorrowing; // All qubits allocated only for borrowing, will be marked with this number or higher.
```

Comment:
Adding more: actually the state is a bit more complicated, since a qubit might be Allocated in one frame and Borrowed in another. I *think* that there are really 3 states: Free, Allocated, and Disabled. The StackFrame class should hold a list of the qubits that have are currently borrowed in that frame. To see if a qubit is currently borrowed, look to see if it's borrowed in the current frame, since that's the only borrowing that matters (if it's borrowed in a higher frame, so what? it can still be borrowed again, and it can't be released, so it's functionally just Allocated).

Reply:
OK, it's even a bit more complicated: if you allocate a qubit just for borrowing, you need to mark it as taken (AllocatedForBorrowing) so you don't return it to an allocation deeper in the stack. That's OK, except that if you allow re-borrowing -- and why not? -- you'd need to keep a "borrowed count" to know when the `Return` should set the state back to Free. I think you can keep this in the borrowed frame by tracking (borrowed qubit, prior state) pairs instead of just borrowed qubits.

Admittedly at this point there's a list of qubit states, a stack or queue of free qubits, and a data structure (probably a dictionary) tracking borrowed qubits in each stack frame, all just to replace one array of longs. Arguably it's not really that much simpler, although at least each data structure has a clear, specific purpose.

Reply:
While I broadly agree that it could be solved differently, I did focus on refactoring the public API rather then rewriting the code from scratch. :) 
I haven't looked close enough to have formed a full judgement on what I think the best internal representation could be that ensures both performance and readability. For this PR, since these are private handles (intentionally), I would like to proceed without doing a full refactoring since the code is in a working state and the priority needs to be to get to a stable API we can commit to.

Reply:
Makes sense to me. We can add this code to the "eventual refactor" list.